### ISSUE TITLE
Type instability with Flux.update! when loss function involves extra arguments

### ISSUE THREAD

**Issue by PatMcFarlane** [FluxML/Flux.jl#2175]
I'm working on replicating an economics application of machine learning that was originally done in Python/TensorFlow [here](https://notes.quantecon.org/submission/5ddb3c926bad3800109084bf).

Essentially the application boils down to solving a functional equation like g(f(x)) - h(f(x)) = 0,  where f(x) does not have a closed form representation and needs to instead be approximated with a neural network.   This requires constructing a loss function where all the data is generated from the neural network itself.  Typically this will necessitate passing parameters into the loss function.

I've found that performance of my replication code is not great.  So I've been checking to see if there are any type instabilities.

I ran into one with ```Flux.update!()```.    Below is a minimal example to illustrate the problem:

```
using Flux

mod1 = Dense(2,1)

expon = 2.0

function DataGen1(model,expon)

    data = ones(2)

    resid = sum(model(data).^expon)

    return resid

end

grads = Flux.gradient(m -> DataGen1(m,expon), mod1)

#check to make sure the gradient is type stable
@code_warntype Flux.gradient(m -> DataGen1(m,expon), mod1)

opt = Flux.setup(Descent(), mod1)

Flux.update!(opt,mod1,grads[1])

@code_warntype Flux.update!(opt,mod1,grads[1])
```

The output from the last ```@code_warntype``` call looked like this:

```
 MethodInstance for Optimisers.update!(::NamedTuple{(:weight, :bias, :σ), Tuple{Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Tuple{}}}, 
::Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}, ::NamedTuple{(:weight, :bias, :σ), Tuple{Matrix{Float32}, Vector{Float32}, Nothing}})
  from update!(tree, model, grad, higher...) in Optimisers at C:\Users\Patrick\.julia\packages\Optimisers\kPdJV\src\interface.jl:65
Arguments
  #self#::Core.Const(Optimisers.update!)
  tree::NamedTuple{(:weight, :bias, :σ), Tuple{Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Tuple{}}}
  model::Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}
  grad::NamedTuple{(:weight, :bias, :σ), Tuple{Matrix{Float32}, Vector{Float32}, Nothing}}
  higher::Tuple{}
Locals
  newmodel::Any
  grads::IdDict{Optimisers.Leaf, Any}
Body::Tuple{NamedTuple{(:weight, :bias, :σ), Tuple{Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Tuple{}}}, Any}
1 ─ %1  = Core.apply_type(Optimisers.IdDict, Optimisers.Leaf, Optimisers.Any)::Core.Const(IdDict{Optimisers.Leaf, Any})
│         (grads = (%1)())
│   %3  = Core.tuple(grads, tree, model, grad)::Tuple{IdDict{Optimisers.Leaf, Any}, NamedTuple{(:weight, :bias, :σ), Tuple{Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Tuple{}}}, Dense{typeof(identity), Matrix{Float32}, Vector{Float32}}, NamedTuple{(:weight, :bias, :σ), Tuple{Matrix{Float32}, Vector{Float32}, Nothing}}}
│         Core._apply_iterate(Base.iterate, Optimisers._grads!, %3, higher)
│   %5  = (:grads, :params)::Core.Const((:grads, :params))
│   %6  = Core.apply_type(Core.NamedTuple, %5)::Core.Const(NamedTuple{(:grads, :params)})
│   %7  = grads::IdDict{Optimisers.Leaf, Any}
│   %8  = Optimisers.IdDict()::IdDict{Any, Any}
│   %9  = Core.tuple(%7, %8)::Tuple{IdDict{Optimisers.Leaf, Any}, IdDict{Any, Any}}
│   %10 = (%6)(%9)::NamedTuple{(:grads, :params), Tuple{IdDict{Optimisers.Leaf, Any}, IdDict{Any, Any}}}
│   %11 = Core.kwfunc(Optimisers._update!)::Core.Const(Optimisers.var"#_update!##kw"())
│         (newmodel = (%11)(%10, Optimisers._update!, tree, model))
│   %13 = Core.tuple(tree, newmodel)::Tuple{NamedTuple{(:weight, :bias, :σ), Tuple{Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Optimisers.Leaf{Optimisers.Descent{Float64}, Nothing}, Tuple{}}}, Any}
└──       return %13
```

If I eliminate the additional argument from the ```DataGen1``` function (e.g., by simply writing ```.^2.0``` instead of ```.^expon```), the problem goes away.  But using literals in this way is very inconvenient for the full-scale application where there are lots of parameters that might need to change frequently.

I'm a novice at Julia, Flux, and machine learning.   So apologies in advance if this is some type of coding error or poor coding practice.

Thanks in advance!

**Comment by ToucheSir** [comment]
> If I eliminate the additional argument from the `DataGen1` function (e.g., by simply writing `.^2.0` instead of `.^expon`), the problem goes away.

Can you post the code and `@code_warntype` output after this change? It shouldn't make a difference. For better or worse, optimization in Flux is inherently type unstable at the moment. We don't deem it a major problem because any performance degradation is minimal (running the optimization procedure takes many orders of magnitudes longer).

**Comment by PatMcFarlane** [comment]
> > If I eliminate the additional argument from the `DataGen1` function (e.g., by simply writing `.^2.0` instead of `.^expon`), the problem goes away.
> 
> Can you post the code and `@code_warntype` output after this change? What you've mentioned shouldn't make a difference. For better or worse, optimization in Flux is inherently type unstable at the moment. We don't deem it a major problem because any performance degradation is minimal (running the optimization procedure takes many orders of magnitudes longer).

Yes you're right.  I checked again and it does type instabilities occur both ways.  Understand the point about updating being a trivial contributor to overall computing time too.

Thanks!
